feat(429 retry after handling): handle 429 retry after from idBroker

UWTech opened the pull request:
add regex to parse API name form idbroker url
add map to store API to retry after time
add methods to check for retry after
add tests

UWTech commented:
@ianwremmel  here are the basics so far.

Cheers.

UWTech commented:
Created interceptor and UTs

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/src/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,128 @@
+   * @param {object} options associated with the request
+   * @returns {number} retry after time in seconds
+   */
+  getRetryAfterTime(options) {
```
let's name this something that doesn't start with get, since it's not a getter for a value related to the interceptor's state

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/src/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,128 @@
+   */
+  setRateLimitExpiry(uri, retryAfter) {
+    const apiName = this.getApiName(uri);
+    const currTimeSec = new Date().getTime() / 1000;
```
Date.now() should be more efficient; note that dividing by 1000 turns the int into a float, so you probably want to wrap the result with `Math.floor`

ianwremmel replied:
(alternatively, it may be easier to work in milliseconds everywhere and just convert the retry-after header from seconds)

UWTech replied:
Done

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/src/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,128 @@
+  getRateLimitStatus(uri) {
+    const apiName = this.getApiName(uri);
+    const currTimeSec = new Date().getTime() / 1000;
+    const dict = rateLimitExpiryTime.get(this);
```
take a look at the TemplateContainer construct (usage of it will probably be more helpful than looking at its definition). it takes care of some of the existence checking for you.

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/src/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,128 @@
+    const apiName = this.getApiName(uri);
+    const currTimeSec = new Date().getTime() / 1000;
+    const dict = rateLimitExpiryTime.get(this);
+    return dict && dict.get(apiName) === null ? false : currTimeSec < dict.get(apiName);
```
This is a bit complicated for a ternary

UWTech replied:
I'm not sure I follow, what is complicated here?

If the value is not in the dict (is null), the result is false, the API is not limited. Else, if the current system time is less than the expiry time, true, we are rate limited. If the currsystem time is not less than the expiry, meaning it is greater than the expiry, the answer is false.

If we didn't have to check for a null value, there wouldn't even need to be a ternary, it would just be return currTimeSec < dict.get(apiName);

I wanted to make sure I wasn't making any assumptions about how null was handled, so rather than doing a direct comparison against a potentially null value, I'm just using a ternary to return false.

UWTech replied:
I did remove the check for a null dict per your earlier comment.

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/src/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,128 @@
+   * @returns {string}
+   */
+  getApiName(uri) {
+    const results = uri === null ? null : uri.match(idBrokerRegex);
```
This is a bit complicated for a ternary

UWTech replied:
This is the use case for ternary. I don't follow what is complicated here.

It has been verified with unit tests.

UWTech replied:
I could skip this ternary and return null, but the second ternary is definitely necessary.

UWTech replied:
There is code that prevents Keying nulls, so shouldn't be a concern.

The code being more compact with ternary is more readable imo, and since they're functionally equivalent,  is it really such a big deal to have ternary?

We're talking about the difference between grepping 7 or 8 lines of code if we use if statements versus 2 lines using ternary operators.

This strikes me as a coding style thing rather than a concern about functional correctness. Is there a style rule saying no ternary operators?

I could add a check in getRateLimitStatus that short circuits and returns if API name is null, but that would be pretty clunky and would obfuscate the method's function imo.  So at line 100 I could add:
if(apiName === null)
 return false;

However, since the expectation is the apiName returned is never null due to the fact we check the Regex against the URI before calling get rate limit status, you are now adding an extra null check on every single positive hit to any idBroker/identity calls. We would expect that apiName is never null, so optimizing with a short circuit for the unexpected case where it is null would be *less* efficient overall.

bzang replied:
Just to chime in here.

I think there is definitely some code style preference here, but also from an architecture standpoint there are some best practices we can use. So, sure there are checks in place higher up the stack that'll short circuit certain behavior, but I think if we're trying to ensure code isolation, then it's important to have checks in place for odd behaviors in case the code calling these Classes get funky. For example, in this implementation, I would think you'd want to have checks in `onResponseError` and `onRequest` for if `options` and `options.uri` are even defined.

I don't personally have an issue with ternaries, and I think if we actually put together a style guide, we'd explicitly encourage them where it makes sense, but compactness doesn't necessarily mean more understandable, and here it definitely took me some extra time to figure out what the ternary is doing.

UWTech replied:
There is no check for options === null in any of the onError interceptor methods.

Provided there is error checking prior to inputting a value into the method at some stage, it should be fine. The most appropriate place to put the logic that does input validation is prior to adding to whatever data structure you want to use to store.

I don't see a legitimate use case for someone accessing the map outside of the accessors provided in the module. Any additional interactions needed should be added after examining the current assumptions and expectations with respect to the map. If people are adding methods, and messing with the map wily-nilly, no amount of preparation is going to prevent problems.

Additionally the module is reasonably small, so it shouldn't be hard for someone to grok the code  and make the appropriate changes.

The discussion around ternary operators is not one of compactness versus excessive catch statements per se; given that the work has been done, and the code is functionally correct, and we agree it is a style preference, I don't think it makes sense to write it again.

I could do an informal survey here at my site, and we'd probably find that most people prefer ternary operators, perhaps at your site, you don't find ternary as easy to understand as lengthy if-then statements and explicit null checks; I don't think one or another is necessarily right or wrong so much as it isn't necessary to reconstitute the solution.

ianwremmel replied:
> given that the work has been done, and the code is functionally correct, and we agree it is a style preference

We don't agree it's *just* "a style preference". Two of the maintainers of this project have raised concerns about readability. We're not saying "don't use ternaries", but we are concerned that these particular ternaries are difficult to grok.

UWTech replied:
Ternary operators can't get much simpler than this. From my perspective you are essentially dictating style.

I'll go ahead and change these, but if you don't want ternary operators being used, you need to have an official rule, and not ask me to change functionally correct code that is already done.

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/src/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,128 @@
+   */
+  isRateLimited(uri) {
+    // determine if the URI is associated with a common identity API
+    if (uri && uri.includes(`idbroker`)) {
```
note that some common identity apis are also hosted on identity.webex.com

UWTech replied:
Thanks, I'll add that.

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/test/unit/spec/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,219 @@
+          }
+
+          interceptor.setRateLimitExpiry(`https://idbroker.webex.com/horse/v1/myID`, 1);
+          sleepForExpiry().then(() => {
```
- promises need to be returned for mocha to wait for them
- .then statements should always be on the following line and indented one level from the above statement (unless the above statement is a `.then` or a `.catch`.

UWTech replied:
Done.

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/test/unit/spec/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,219 @@
+            return new Promise((resolve) => setTimeout(resolve, ms));
+          }
+
+          async function sleepForExpiry() {
```
you've got a few different sleep definitions in this file, how about consolidating them at the top as one function.

UWTech replied:
Done.

Not sure how I ended up with the replication and two different methods.

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/test/unit/spec/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,219 @@
+          }
+
+          async function sleepForExpiry() {
+            await sleep(2000);
```
Awesome as async/await is, we're keeping it out of source and test code for now.

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/src/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,128 @@
+    const currTimeSec = new Date().getTime() / 1000;
+    const expiry = currTimeSec + retryAfter;
+    const dict = rateLimitExpiryTime.get(this);
+    return apiName && dict && dict.set(apiName, expiry);
```
Doesn't checking of `dict` is valid hide a real error?

ianwremmel replied:
I think you'll also want to use safeSetTimeout to setup cache eviction, right? Otherwise, especially as this expands to other services, you'll end up with a memory leak.

UWTech replied:
How would I end up with a memory leak?

These are all keyed by API name, so any new entries would overwrite the old entry.

For example, dostuff/1234 and dostuff/4567 map to the same key "dostuff". I don't see why we need cache eviction for an ephemeral structure that is gone after the app shuts down. 

I can remove the check for dict, I added it because you requested it be put in in the examples provided.

UWTech commented:
Addressed code review feedback

UWTech commented on packages/node_modules/@ciscospark/spark-core/src/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,130 @@
+   */
+  setRateLimitExpiry(uri, retryAfter) {
+    const apiName = this.getApiName(uri);
+    const currTimeSec = new Date().getTime();
```
The change was made, but the variable name was not updated to milli ( I changed the method docs to reflect this, but didn't update the variable name.)

I can get that real quick.

UWTech commented on packages/node_modules/@ciscospark/spark-core/src/index.js:
```diff
@@ -42,6 +42,7 @@ export {default as RequestLoggerInterceptor} from './interceptors/request-logger
 export {default as RequestTimingInterceptor} from './interceptors/request-timing';
 export {default as SparkTrackingIdInterceptor} from './interceptors/spark-tracking-id';
 export {default as SparkUserAgentInterceptor} from './interceptors/spark-user-agent';
+export {default as RateLimitInterceptor} from './interceptors/rate-limit';
```
I'll take a look, thanks.

UWTech replied:
Done.

ianwremmel replied:
I still don't see a change that adds the rate limit interceptor to the list of active interceptors.

UWTech replied:
Right I hadn't pushed the commit since there was still disagreement on the use of ternary operators. I have added it to spark-js-sdk/packages/node_modules/@ciscospark/spark-core/src/spark-core.js, so should show up when I push.

UWTech commented:
@ianwremmel  any further concerns here?

Getting pinged asking when this is getting merged in.

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/src/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,130 @@
+    const apiName = this.getApiName(uri);
+    const currTimeMilli = new Date().getTime();
+    const dict = rateLimitExpiryTime.get(this);
+    return dict.get(apiName) === null ? false : currTimeMilli < dict.get(apiName);
```
when would `dict.get(apiName) === null` be true? WeakMaps return `undefined` if the specified key can't be found.

UWTech replied:
Ah, gotcha. I'll change that to undefined, thanks Ian.

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/src/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,130 @@
+    const currTimeMilli = new Date().getTime();
+    const expiry = currTimeMilli + retryAfter;
+    const dict = rateLimitExpiryTime.get(this);
+    return apiName && dict.set(apiName, expiry);
```
Wouldn't the `apiName` check make more sense around line 87 to avoid potentially doing needless work if `apiName` is falsy?

UWTech replied:
Changed.

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/src/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,130 @@
+  onResponseError(options, reason) {
+    if (reason.statusCode === 429 && (options.uri.includes(`idbroker`) || options.uri.includes(`identity`))) {
+      // set the retry after in the map, setting to milliseconds
+      this.setRateLimitExpiry(options.uri, this.extractRetryAfterTime(options) * 1000);
```
The `*1000` might make more sense inside `extractRetryAfterTime`; that way, all functions' inputs and outputs are in the same unit.

UWTech replied:
I don't really like the idea of the method mutating the form of the data which is used as input, but I don't have strong objections. Consistency is important as well, so I can make that change.

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/test/unit/spec/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,291 @@
+        });
+        it(`returns false when previously rate limited idbroker API expiry time has been met`, () => {
+          interceptor.setRateLimitExpiry(`https://idbroker.webex.com/horse/v1/myID`, 1);
+          return sleep(2000)
```
Take a look at our use of `lolyx` for mock timers so that tests can be written synchronously (and run significantly faster).

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/test/unit/spec/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,291 @@
+          assert.equal(interceptor.getApiName(null), null);
+        });
+        it(`returns idbroker API name if there is a match`, () => {
+          assert.equal(interceptor.getApiName(`https://idbroker.webex.com/horse/v1/myID`), `horse`);
```
Is the goal to stop using an entire hostname when we get back a 429 or to stop using the particular endpoint that sent the 429?

UWTech replied:
As discussed in earlier commit messages, the specific API associated with a hostname is limited. The two host names that we've been asked to honor 429s from are the ID broker endpoints.

ianwremmel replied:
gotcha. sorry, that still seems weird to me, but if that's the plan of record, good enough.

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/test/unit/spec/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,291 @@
+
+      describe(`#getRateLimitStatus`, () => {
+        it(`returns false if API name is not in rate limit expiry map`, () => {
+          assert.equal(interceptor.getRateLimitStatus(`https://fakeuir.com/testFake`), false);
```
Let's use `example.com` for all domains we don't control. Yes, these are unit tests, but we're using a lot of real domains for testing throughout the product that have definitely caused wierd issues.

ianwremmel commented on packages/node_modules/@ciscospark/spark-core/test/unit/spec/interceptors/rate-limit.js:
```diff
@@ -0,0 +1,291 @@
+      });
+
+      describe(`#onRequest`, () => {
+        it(`Rejects idbroker request if API is rate limited`, () => {
```
See question above about rate-limiting a service vs an endpoint

UWTech replied:
Answered earlier.

UWTech commented:
@Ian got rid of the ternary operators; you're making an old embedded systems developer sad, but I can avoid their use in the future if it is going to be so controversial. I honestly find them much more readable, and when they are not nested to an insane degree, I prefer them.

Updated to use lolex, moved some stuff around, and I did add the interceptor in the core.

UWTech commented:
@ianwremmel Tagged the wrong Ian on last comment....

got rid of the ternary operators; you're making an old embedded systems developer sad, but I can avoid their use in the future if it is going to be so controversial. I honestly find them much more readable, and when they are not nested to an insane degree, I prefer them.

Updated to use lolex, moved some stuff around, and I did add the interceptor in the core.

Cheers.

ianwremmel commented:
already merged

UWTech commented:
Merged. Sorry about skipping cleanup here.
